File: karaoke/services/song_config_service.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from tortoise.exceptions import DoesNotExist

from karaoke.audio_layout import merge_manual_roles, parse_audio_layout, serialize_audio_layout
from karaoke.domain.playback import has_full_override, refresh_playback_mode, resolve
from karaoke.domain.prepare_policy import profile_needs_prepare
from karaoke.dto.mappers import playback_detail, song_item
from karaoke.embedded import probe_and_save_layout
from karaoke.errors import fail_result
from karaoke.infra.repositories.history_repo import HistoryRepository
from karaoke.infra.repositories.song_repo import SongRepository
from karaoke.results import Result
from karaoke.services.prepare_service import PrepareService


class SongConfigService:
    def __init__(
        self,
        songs: SongRepository = None,
        histories: HistoryRepository = None,
        prepare: PrepareService = None,
    ) -> None:
        self._songs = songs or SongRepository()
        self._histories = histories or HistoryRepository()
        self._prepare = prepare or PrepareService()

    async def get_detail(self, song_id: int) -> Result:
        result = Result()
        try:
            song = await self._songs.get(song_id)
            profile = resolve(song)
            result.data = {
                'id': song.id,
                'display_name': song.display_name,
                'source_path': song.source_path,
                'source_origin': song.source_origin,
                'source_rel': song.source_rel,
                'is_playable': song.is_playable,
                **playback_detail(song, profile),
            }
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不存在"
        except Exception as exc:
            fail_result(result, exc, "获取歌曲详情失败")
        return result

    async def patch(self, song_id: int, body: dict) -> Result:
        result = Result()
        try:
            song = await self._songs.get(song_id)
            display_name = body.get('display_name')
            audio_tracks = body.get('audio_tracks')
            if display_name:
                song.display_name = display_name.strip()[:256]
                await self._songs.save(song, ['display_name', 'update_time'])
                for h in await self._histories.list_for_song(song.id):
                    h.name = song.display_name
                    await self._histories.save(h, ['name', 'update_time'])
            if audio_tracks is not None:
                layout = merge_manual_roles(parse_audio_layout(song.audio_layout), audio_tracks)
                song.audio_layout = serialize_audio_layout(layout)
                await self._songs.save(song, ['audio_layout', 'update_time'])
            profile = await refresh_playback_mode(song)
            result.data = song_item(song, profile)
            result.msg = "更新成功"
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不存在"
        except Exception as exc:
            fail_result(result, exc, "更新歌曲失败")
        return result

    async def detect_playback(self, song_id: int) -> Result:
        result = Result()
        try:
            song = await self._songs.get(song_id)
            if not has_full_override(song.display_name)[0]:
                await probe_and_save_layout(song, assigned_by='auto')
            profile = await refresh_playback_mode(song)
            result.data = playback_detail(song, profile)
            if profile_needs_prepare(song, profile):
                prep = await self._prepare.schedule(song_id)
                result.data = {**result.data, 'prepare': prep}
            result.msg = "播放能力检测完成"
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不存在"
        except Exception as exc:
            fail_result(result, exc, "检测播放能力失败")
        return result

    async def request_prepare(self, song_id: int, wait: bool = False) -> Result:
        result = Result()
        try:
            song = await self._songs.get(song_id)
            if has_full_override(song.display_name)[0]:
                profile = resolve(song)
                result.data = playback_detail(song, profile)
                result.msg = "已有 __override__ 三件套，无需预生成内嵌缓存"
                return result
            prep = await self._prepare.schedule(song_id)
            if wait:
                prep = await self._prepare.wait_until_ready(song_id)
            profile = await refresh_playback_mode(song)
            result.data = {
                **playback_detail(song, profile),
                'prepare': prep,
                'cache_ready': prep.get('ready', False),
            }
            if prep.get('ready'):
                result.msg = "缓存已就绪"
            elif prep.get('status') in ('pending', 'running'):
                result.msg = "正在后台生成缓存"
            else:
                result.code = 1
                result.msg = prep.get('error') or "缓存生成失败"
        except DoesNotExist:
            result.code = 1
            result.msg = "歌曲不存在"
        except Exception as exc:
            fail_result(result, exc, "预生成缓存失败")
        return result

File: crates/services/src/song_config_service.rs
//! 歌曲配置服务：详情/编辑/播放能力检测/预生成缓存。
//! 对应 Python `karaoke/services/song_config_service.py`。

use crate::dto::{db_error_message, ApiResult};
use crate::mappers::{playback_detail, song_item};
use karaoke_domain::audio_layout::{merge_manual_roles, TrackRole};
use karaoke_domain::playback::effective_mode;
use karaoke_infra::embedded::probe_layout;
use karaoke_infra::repositories::{HistoryRepository, SongRepository};
use karaoke_infra::PlaybackResolver;
use karaoke_jobs::PrepareTaskManager;
use serde_json::Value;
use std::sync::Arc;
use std::time::Duration;

#[derive(Clone)]
pub struct SongConfigService {
    pub songs: SongRepository,
    pub histories: HistoryRepository,
    pub resolver: PlaybackResolver,
    pub prepare: Arc<PrepareTaskManager>,
}

impl SongConfigService {
    pub async fn get_detail(&self, song_id: i64) -> ApiResult {
        let song = match self.songs.get_optional(song_id).await {
            Ok(Some(s)) => s,
            Ok(None) => return ApiResult::fail("歌曲不存在"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "获取歌曲详情失败")),
        };
        let profile = self.resolver.resolve(&song).await;
        let override_status = self.resolver.override_file_status(&song.display_name).await;
        let detail = playback_detail(&song, &profile, override_status);
        let mut value = serde_json::to_value(detail).unwrap_or(Value::Null);
        if let Value::Object(map) = &mut value {
            map.insert("id".to_string(), serde_json::json!(song.id));
            map.insert(
                "display_name".to_string(),
                serde_json::json!(song.display_name),
            );
            map.insert(
                "source_path".to_string(),
                serde_json::json!(song.source_path),
            );
            map.insert(
                "source_origin".to_string(),
                serde_json::json!(song.source_origin),
            );
            map.insert("source_rel".to_string(), serde_json::json!(song.source_rel));
            map.insert(
                "is_playable".to_string(),
                serde_json::json!(song.is_playable),
            );
        }
        ApiResult::ok_with_data(value)
    }

    pub async fn patch(&self, song_id: i64, body: &Value) -> ApiResult {
        let song = match self.songs.get_optional(song_id).await {
            Ok(Some(s)) => s,
            Ok(None) => return ApiResult::fail("歌曲不存在"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "更新歌曲失败")),
        };

        if let Some(name) = body.get("display_name").and_then(Value::as_str) {
            let trimmed: String = name.trim().chars().take(256).collect();
            if !trimmed.is_empty() {
                if let Err(e) = self.songs.update_display_name(song_id, &trimmed).await {
                    return ApiResult::fail(db_error_message(&e, "更新歌曲失败"));
                }
                for h in self
                    .histories
                    .list_for_song(song_id)
                    .await
                    .unwrap_or_default()
                {
                    let _ = self.histories.update_name(h.id, &trimmed).await;
                }
            }
        }

        if let Some(tracks) = body.get("audio_tracks").and_then(Value::as_array) {
            let updates: Vec<(i32, TrackRole)> = tracks
                .iter()
                .filter_map(|item| {
                    let index = item.get("index").and_then(Value::as_i64)? as i32;
                    let role = item
                        .get("role")
                        .and_then(Value::as_str)
                        .and_then(TrackRole::parse)?;
                    Some((index, role))
                })
                .collect();
            let merged = merge_manual_roles(song.layout(), &updates);
            if let Err(e) = self.songs.update_audio_layout(song_id, &merged).await {
                return ApiResult::fail(db_error_message(&e, "更新歌曲失败"));
            }
        }

        let song = match self.songs.get(song_id).await {
            Ok(s) => s,
            Err(e) => return ApiResult::fail(db_error_message(&e, "更新歌曲失败")),
        };
        let profile = self.resolver.resolve(&song).await;
        if let Err(e) = self
            .songs
            .update_playback_meta(
                song_id,
                effective_mode(&profile, &song.playback_mode).as_str(),
                Some(profile.playback_source.as_str()),
                profile.can_queue,
            )
            .await
        {
            tracing::warn!("persist playback meta failed for song {song_id}: {e}");
        }
        let song = self.songs.get(song_id).await.unwrap_or(song);
        ApiResult::ok_msg_data("更新成功", song_item(&song, Some(&profile), None))
    }

    pub async fn detect_playback(&self, song_id: i64) -> ApiResult {
        let song = match self.songs.get_optional(song_id).await {
            Ok(Some(s)) => s,
            Ok(None) => return ApiResult::fail("歌曲不存在"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "检测播放能力失败")),
        };

        if !self.resolver.override_complete(&song.display_name).await {
            let layout = probe_layout(&self.resolver.media, &song.source_path, "auto").await;
            if let Err(e) = self.songs.update_audio_layout(song_id, &layout).await {
                tracing::warn!("persist audio_layout failed for song {song_id}: {e}");
            }
        }

        let song = self.songs.get(song_id).await.unwrap_or(song);
        let profile = self.resolver.resolve(&song).await;
        if let Err(e) = self
            .songs
            .update_playback_meta(
                song_id,
                effective_mode(&profile, &song.playback_mode).as_str(),
                Some(profile.playback_source.as_str()),
                profile.can_queue,
            )
            .await
        {
            tracing::warn!("persist playback meta failed for song {song_id}: {e}");
        }
        let song = self.songs.get(song_id).await.unwrap_or(song);

        let override_status = self.resolver.override_file_status(&song.display_name).await;
        let detail = playback_detail(&song, &profile, override_status);
        let mut data = serde_json::to_value(&detail).unwrap_or(Value::Null);

        if self.resolver.needs_prepare(&song, &profile).await {
            let prep = self.prepare.schedule(song_id).await;
            if let Value::Object(map) = &mut data {
                map.insert(
                    "prepare".to_string(),
                    serde_json::to_value(prep).unwrap_or(Value::Null),
                );
            }
        }
        ApiResult::ok_msg_data("播放能力检测完成", data)
    }

    pub async fn request_prepare(&self, song_id: i64, wait: bool) -> ApiResult {
        let song = match self.songs.get_optional(song_id).await {
            Ok(Some(s)) => s,
            Ok(None) => return ApiResult::fail("歌曲不存在"),
            Err(e) => return ApiResult::fail(db_error_message(&e, "预生成缓存失败")),
        };

        if self.resolver.override_complete(&song.display_name).await {
            let profile = self.resolver.resolve(&song).await;
            let override_status = self.resolver.override_file_status(&song.display_name).await;
            let detail = playback_detail(&song, &profile, override_status);
            return ApiResult::ok_msg_data("已有 __override__ 三件套，无需预生成内嵌缓存", detail);
        }

        let mut prep = self.prepare.schedule(song_id).await;
        if wait {
            prep = self
                .prepare
                .wait_until_ready(song_id, Duration::from_secs(3600))
                .await;
        }

        let song = self.songs.get(song_id).await.unwrap_or(song);
        let profile = self.resolver.resolve(&song).await;
        if let Err(e) = self
            .songs
            .update_playback_meta(
                song_id,
                effective_mode(&profile, &song.playback_mode).as_str(),
                Some(profile.playback_source.as_str()),
                profile.can_queue,
            )
            .await
        {
            tracing::warn!("persist playback meta failed for song {song_id}: {e}");
        }

        let override_status = self.resolver.override_file_status(&song.display_name).await;
        let detail = playback_detail(&song, &profile, override_status);
        let mut data = serde_json::to_value(&detail).unwrap_or(Value::Null);
        if let Value::Object(map) = &mut data {
            map.insert(
                "prepare".to_string(),
                serde_json::to_value(&prep).unwrap_or(Value::Null),
            );
            map.insert("cache_ready".to_string(), serde_json::json!(prep.ready));
        }

        if prep.ready {
            ApiResult::ok_msg_data("缓存已就绪", data)
        } else if prep.status == "pending" || prep.status == "running" {
            ApiResult::ok_msg_data("正在后台生成缓存", data)
        } else {
            let msg = prep
                .error
                .clone()
                .unwrap_or_else(|| "缓存生成失败".to_string());
            ApiResult::fail_with_data(msg, data)
        }
    }
}
